blueocean-dashboard/src/main/js/util/smart-fetch.js

```diff
@@ -0,0 +1,277 @@
+import isoFetch from 'isomorphic-fetch';
+
+/**
+ * Freezes an object and all child properties
+ */
+const deepFreeze = (obj) => {
+    const propNames = Object.getOwnPropertyNames(obj);
+    for (let idx = 0; idx < propNames.length; idx++) {
+        const prop = obj[propNames[idx]];
+        if (prop !== null) {
+            if (typeof prop === 'object') {
+                deepFreeze(prop);
+            }
+        }
+    }
+    return Object.freeze(obj);
+};
+
+/**
+ * Validates the status is 200-299 or returns an error
+ */
+const checkStatus = (response) => {
+    if (response.status >= 300 || response.status < 200) {
+        console.error('ERROR: ', response); // eslint-disable-line no-console
+        const error = new Error(response.statusText);
+        error.response = response;
+        throw error;
+    }
+    return response;
+};
+
+/**
+ * Get the JSON response body
+ */
+const parseJSON = (rsp) => {
+    try {
+        return rsp.json();
+    } catch (err) {
+        console.error('Unable to parse JSON: ', rsp.body, err); // eslint-disable-line no-console
+        throw new Error('Invalid JSON payload', err);
+    }
+};
+
+/**
+ * Static fetch options used for every request
+ */
+const fetchOptions = { credentials: 'same-origin' };
+
+/**
+ * DuplicateCallTracker maintains active fetches against a particular key
+ */
+export class DuplicateCallTracker {
+    constructor() {
+        /**
+         * Onload callbacks cache. Used to ensure we don't
+         * issue multiple in-parallel requests for the same
+         * class metadata.
+         */
+        this.fetchCallbacks = {};
+        this.promises = {};
+    }
+
+    /**
+     * Generalization of consolidation of duplicate requests
+     */
+    execute(key, promiseProvider, onComplete) {
+        let callbacks = this.fetchCallbacks[key];
+        if (!callbacks) {
+            // This is the first request for this type. Initialize the
+            // callback cache and then issue the request
+            callbacks = this.fetchCallbacks[key] = onComplete ? [onComplete] : [];
+
+            promiseProvider()
+                .then((data) => {
+                    // Notify all callbacks
+                    for (let i = 0; i < callbacks.length; i++) {
+                        try {
+                            callbacks[i](data);
+                        } catch (e) {
+                            console.error('Unexpected Error fulfilling callbacks', e); // eslint-disable-line no-console
+                        }
+                    }
+                    delete this.fetchCallbacks[key];
+                    return data; // for downstream .then
+                });
+        } else {
+            // We already have an in-flight request, just store the onComplete callback
+            callbacks.push(onComplete);
+        }
+    }
+
+
+    /**
+     * Generalization of consolidation of duplicate requests
```

Comment:
"Generalization of consolidation of duplicate requests" ... duplicated on this function and the previous. In both cases, it doesn't really help me understand what the functions are really doing. Sounds good though :)